src/history/hs_rec.c

```diff
@@ -1046,3 +1046,103 @@ __hs_delete_reinsert_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, ui
 
     return (ret);
 }
+
+/*
+ * __hs_delete_record --
+ *     Delete the update left in the history store
+ */
+static int
+__hs_delete_record(
+  WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_ITEM *key, WT_UPDATE *upd, WT_UPDATE *tombstone)
+{
+    WT_DECL_RET;
+    bool hs_read_committed;
+#ifdef HAVE_DIAGNOSTIC
+    WT_TIME_WINDOW *hs_tw;
+#endif
+
+    if (r->hs_cursor == NULL)
+        WT_RET(__wt_curhs_open(session, NULL, &r->hs_cursor));
+    hs_read_committed = F_ISSET(r->hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
+    F_SET(r->hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
+
+    /* No need to delete from the history store if it is already obsolete. */
+    if (tombstone != NULL && __wt_txn_upd_visible_all(session, tombstone))
+        goto done;
+
+    r->hs_cursor->set_key(r->hs_cursor, 4, S2BT(session)->id, key, WT_TS_MAX, UINT64_MAX);
+    WT_ERR_NOTFOUND_OK(__wt_curhs_search_near_before(session, r->hs_cursor), true);
+    /* It's possible the value in the history store becomes obsolete concurrently. */
+    if (ret == WT_NOTFOUND) {
+        WT_ASSERT(session, tombstone != NULL && __wt_txn_upd_visible_all(session, tombstone));
+        ret = 0;
+        goto done;
+    }
+
+#ifdef HAVE_DIAGNOSTIC
+    __wt_hs_upd_time_window(r->hs_cursor, &hs_tw);
+    WT_ASSERT(session, hs_tw->start_txn == WT_TXN_NONE || hs_tw->start_txn == upd->txnid);
+    WT_ASSERT(session, hs_tw->start_ts == WT_TS_NONE || hs_tw->start_ts == upd->start_ts);
+    WT_ASSERT(
+      session, hs_tw->durable_start_ts == WT_TS_NONE || hs_tw->durable_start_ts == upd->durable_ts);
+    if (tombstone != NULL) {
+        WT_ASSERT(session, hs_tw->stop_txn == tombstone->txnid);
+        WT_ASSERT(session, hs_tw->stop_ts == tombstone->start_ts);
+        WT_ASSERT(session, hs_tw->durable_stop_ts == tombstone->durable_ts);
+    } else
+        WT_ASSERT(session, !WT_TIME_WINDOW_HAS_STOP(hs_tw));
+#endif
+
+    WT_ERR(r->hs_cursor->remove(r->hs_cursor));
+done:
+    if (tombstone != NULL)
+        F_CLR(tombstone, WT_UPDATE_TO_DELETE_FROM_HS | WT_UPDATE_HS);
+    F_CLR(upd, WT_UPDATE_TO_DELETE_FROM_HS | WT_UPDATE_HS);
+
+err:
+    if (!hs_read_committed)
+        F_CLR(r->hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
```

Comment:
In the case of an error, we would jump to the label `err` and clear the flag of `hs_cursor` if required. Why wouldn't we clear the flag(s) for the `tombstone` and `update` as well?
Instead of:
```
WT_ERR(r->hs_cursor->remove(r->hs_cursor));
```
We could simply do:
```
ret = r->hs_cursor->remove(r->hs_cursor);
```
What do you think?

Reply:
If we failed to remove, then the update is still in the history store, we cannot clear the flag in this case. We should retry to remove again next time.